packages/studio-ui/src/web/components/Layout/TopNav/DeployCloudBtn.tsx

```diff
@@ -0,0 +1,111 @@
+import { Icon, Tooltip, Spinner } from '@blueprintjs/core'
+import axios from 'axios'
+import { lang, toast } from 'botpress/shared'
+import classNames from 'classnames'
+import { Training } from 'common/nlu-training'
+import moment from 'moment'
+import React, { useCallback, useEffect, useState } from 'react'
+import { connect } from 'react-redux'
+import { RootReducer } from '~/reducers'
+
+import style from './style.scss'
+
+interface Props {
+  trainSession: Training
+}
+
+const NeedTraining = () => {
+  return <div className={style.tooltip}>{lang.tr('topNav.deploy.tooltip.needsTraining')}</div>
+}
+const DeployInfo = ({ lastImportDate }) => {
+  const importDateStr = useCallback(() => {
+    if (!lastImportDate) {
+      return null
+    }
+
+    return moment(lastImportDate).fromNow()
+  }, [lastImportDate])
+
+  return (
+    <div className={style.tooltip}>
+      {lastImportDate
+        ? [lang.tr('topNav.deploy.tooltip.lastDeploy'), importDateStr()].join(' ')
+        : lang.tr('topNav.deploy.tooltip.deployBot')}
+    </div>
+  )
+}
+const Deploying = () => {
+  return <div className={style.tooltip}>{lang.tr('topNav.deploy.tooltip.deploying')}</div>
+}
+
+const DeployCloudBtn = (props: Props) => {
+  const { trainSession } = props
+  const [lastImportDate, setLastImportDate] = useState(null)
+  const [deployLoading, setDeployLoading] = useState(false)
+
+  useEffect(() => {
+    axios
+      .get(`${window.STUDIO_API_PATH}/cloud/info`)
+      .then((res) => res.data)
+      .then((data) => {
+        const { imported_on } = data
+        setLastImportDate(imported_on)
+      })
+      .catch((e) => {})
+  }, [])
+
+  const isTrained = useCallback(() => {
+    return Object.keys(trainSession).reduce((trained, lang) => {
+      if (trainSession[lang].status !== 'done') {
+        trained = false
+      }
+      return trained
+    }, true)
+  }, [trainSession])
+
+  const deployToCloud = useCallback(() => {
+    if (deployLoading || !isTrained()) {
+      return
+    }
+    toast.info(lang.tr('topNav.deploy.toaster.info'))
+    setDeployLoading(true)
+    axios
+      .get(`${window.STUDIO_API_PATH}/cloud/deploy`)
+      .then((res) => res.data)
+      .then((data) => {
+        const { imported_on } = data
+        setTimeout(() => {
+          toast.success(lang.tr('topNav.deploy.toaster.success'))
+          setLastImportDate(imported_on)
+          setDeployLoading(false)
+        }, 10000)
```

Comment:
the deployment speed is very quick, it can seem like nothing happened. In order to make the button click feel like its doing something and to avoid spam clicking there is a timeout for the use of the button. 10 seconds felt right.

Reply:
But what if the import, for some reason, takes more than 10sec? I think it can happen if you have a slow connection and try to upload an archive of ~100mb.

Reply:
From what I understand, you deploy the bot, then 10 seconds later you show a toast saying "Deploy complete" and re-enable the button. Waiting 10 seconds even if the bot took 1 sec to deploy. 

The button should be disabled only until the deploy is complete, after that, it should be clickable again. As Laurent said, some deploy might take more than 10 seconds. Having fixed timeouts may feel weird in some cases

Reply:
**suggestion**: Just going to chime in here, feel free to ignore; I have multiple examples of this concept in the `cdm-cloud` frontend.

Having local state to handle the loading is good, so you can `setLoader(true) before launching your promise, considering all validations passed, and then in a `.finally()` chain reset the loader to false. This means the loader will only appear for the duration of the call, which makes sense.

Because this sometimes only flicker and is not perceptible to the human eye, what I've done is have a toaster for displaying to users the success or failure state of their operation.